### PR TITLE
[Feature/OTX] Common & Detection Task Refactoring to otx.algorithms

### PR DESCRIPTION
**Pull request by harimkang**
## Summary
Refactor external backends to otx algorithms of Base (common) & Detection

### This PR includes :
- Add otx.algorithms.common for common base code of classification / detection / segmentation tasks
- Add ote.algorithms.detection for Detection task (mmdetection + model-preparation-algorithm)
- Detection NNCF & POT Task refactoring
- Add common & detection adapters
- (TODO) Installation for training -> Next PR
- (TODO) Test code Refactoring for detection -> Next PR

### Test command (Sample Code for detection training)
```
bash otx/algorithms/detection/init_venv.sh ./venv python3.8
source venv/bin/activate
python -m otx.algorithms.detection.tools.detection_sample otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml --export
```

### CODE REVIEW

File: otx/algorithms/detection/tasks/dataset/__init__.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+"""OTX Algorithms - Detection Dataset."""
```
Comment (supersoob) [inline otx/algorithms/detection/tasks/dataset/__init__.py]: I don't know if it's proper to put dataset-related codes in tasks folder. In anomaly, task folder is for such type as train/inference/nncf. some modules like model, datasets were in adapter folder. In my case, I made new subfolder called extension like `otx/algorithms/classification/extension`.

Comment (harimkang) [reply otx/algorithms/detection/tasks/dataset/__init__.py]: I agree that it is not the right path. It can probably be moved to the adapters folder. (adapter is not ready now) 

and I think a new module called extension is not good idea. i think we can use adapters submodule.

Comment (supersoob) [reply otx/algorithms/detection/tasks/dataset/__init__.py]: Okay. Since we are just in initial refactoring phase, let's take dataset in auxiliary path and move on to adapter folder.

Comment (harimkang) [reply otx/algorithms/detection/tasks/dataset/__init__.py]: '+ Maybe the codes included in utils can be modified to belong to adapters. (If the adapters are implemented, I will consider it.)

File: tox.ini [hunk 2]
```diff
@@ -23,7 +23,10 @@ deps =
     -r{toxinidir}/requirements/base.txt
     -r{toxinidir}/requirements/anomaly.txt
     -r{toxinidir}/requirements/openvino.txt
-commands = pre-commit run --all-files
+    -r{toxinidir}/requirements/common.txt
+commands = 
+    pip install --no-cache-dir --no-binary=mmpycocotools mmpycocotools
```
Comment (ashwinvaidya17) [inline tox.ini]: What's the advantage of `--no-cache-dir`? Wouldn't it slow the CI if we download and install it each time?

File: otx/algorithms/common/adapters/__init__.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+"""Adapters for OTX Common Algorithms."""
+
+# Copyright (C) 2022 Intel Corporation
```
Comment (ashwinvaidya17) [inline otx/algorithms/common/adapters/__init__.py]: Minor issue but for consistency we can stick to spdx license header

File: requirements/common.txt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+# Common Tasks Requirements.                                                       #
+-e otx/algorithms/common/adapters/mpa
```
Comment (ashwinvaidya17) [inline requirements/common.txt]: Why is this a separate requirement and why does it point to an internal folder? I think a similar thing can be achieved from setup.py https://docs.python.org/3/distutils/examples.html?highlight=package_dir#pure-python-distribution-by-package. But I am not clear with this design. If we want the stuff in mpa to be accessable from `otx` root then we should probably restructure our repository.

File: otx/algorithms/common/configs/base.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
+"""Base Configuration of OTX Common Algorithms."""
```
Comment (ashwinvaidya17) [inline otx/algorithms/common/configs/base.py]: Is this a base configuration for all the models? What if there are models which do not require training (like anomaly)? Should we move it to `base/configs/training` or something similar that implies that models that require training should import this config.

File: otx/algorithms/common/utils/hooks.py [hunk 6]
```diff
@@ -0,0 +1,569 @@
+"""Collections of hooks for common OTX algorithms."""
+
```
Comment (ashwinvaidya17) [inline otx/algorithms/common/utils/hooks.py]: These hooks seem mmx specific. Should we move it to `common/utils/mmx` or something similar?

File: otx/algorithms/common/utils/runner.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
+"""Runner with cancel for common OTX algorithms."""
```
Comment (ashwinvaidya17) [inline otx/algorithms/common/utils/runner.py]: Same here

File: otx/algorithms/detection/tasks/dataset/__init__.py (hunk 1 above)
Comment (ashwinvaidya17) [reply otx/algorithms/detection/tasks/dataset/__init__.py]: I agree. Adapters will be a better location

File: otx/algorithms/detection/tasks/nncf.py [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+            task_type_to_label_domain(self._task_type),
+            random_seed=42,
+        )
+        set_hyperparams(self._config, self._hyperparams)
+        self.confidence_threshold: float = self._hyperparams.postprocessing.confidence_threshold
+
+        # Set default model attributes.
+        self._optimization_methods = []  # type: List
+        self._precision = self._precision_from_config
+        self._optimization_type = ModelOptimizationType.MO
```
Comment (ashwinvaidya17) [inline otx/algorithms/detection/tasks/nncf.py]: Isn't `optimization_type` being overwritter on line 157?

File: otx/algorithms/detection/tasks/nncf.py [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+                        get_fake_input_func=get_fake_input,
+                    )
+                    logger.info("Loaded model weights from Task Environment and wrapped by NNCF")
+                else:
+                    try:
+                        load_state_dict(model, model_data["model"])
+
+                        # It prevent model from being overwritten
+                        # if "load_from" in self._config:
+                        #    self._config.load_from = None
```
Comment (ashwinvaidya17) [inline otx/algorithms/detection/tasks/nncf.py]: Minor issue but if these lines are not needed then we can remove them.

**Review by ashwinvaidya17 — COMMENTED** [review]
Thanks. I have a few comments.

**Comment by harimkang** [comment]
> Thanks. I have a few comments.

Hi @ashwinvaidya17 ,
Thanks for the review!
I'm still thinking about the mpa (& each algo) installation and this PR still has work to do. I will continue to work with reference to the comments. Also, when this PR is ready, I will request a review again. :)

File: otx/algorithms/detection/tasks/dataset/__init__.py (hunk 1 above)
Comment (harimkang) [reply otx/algorithms/detection/tasks/dataset/__init__.py]: @supersoob Could you please check this one more time?

File: tox.ini (hunk 2 above)
Comment (harimkang) [reply tox.ini]: Removed this.

File: otx/algorithms/common/adapters/__init__.py (hunk 3 above)
Comment (harimkang) [reply otx/algorithms/common/adapters/__init__.py]: The header was constructed by referring to anomaly as much as possible. Could you check it one more time?

File: requirements/common.txt (hunk 4 above)
Comment (harimkang) [reply requirements/common.txt]: There are still a lot of problems with installation (related to lib version). In this PR, only the necessary modifications are made in tox.ini (Not requirements/) so i removed this file

File: otx/algorithms/detection/adapters/mmdet/dataset/mmdataset.py [hunk 10]
```diff
@@ -0,0 +1,238 @@
+"""Base MMDataset for Detection Task."""
```
Comment (supersoob) [inline otx/algorithms/detection/adapters/mmdet/dataset/mmdataset.py]: I'd like to suggest to change this file name to otx_det_dataset.py. What do you think?

File: otx/algorithms/common/configs/base.py (hunk 5 above)
Comment (harimkang) [reply otx/algorithms/common/configs/base.py]: I have reviewed your suggestions. Currently, everything in `otx/algorithms/common` is for the training backend (These are files commonly used in classification/detection/segmentation training. not anomaly). 

The name `common` doesn't seem appropriate, do you have any other ideas?

File: otx/algorithms/common/utils/hooks.py (hunk 6 above)
Comment (harimkang) [reply otx/algorithms/common/utils/hooks.py]: Thanks for the nice comment. I modified it in a similar way, please check it again.

File: otx/algorithms/detection/tasks/nncf.py (hunk 8 above)
Comment (harimkang) [reply otx/algorithms/detection/tasks/nncf.py]: Good point. thanks! This has been corrected.

File: otx/algorithms/detection/tasks/nncf.py (hunk 9 above)
Comment (harimkang) [reply otx/algorithms/detection/tasks/nncf.py]: Yes I deleted this.

File: otx/algorithms/detection/adapters/mmdet/config_utils.py [hunk 11]
```diff
@@ -0,0 +1,394 @@
+"""Collection of utils for task implementation in Detection Task."""
```
Comment (supersoob) [inline otx/algorithms/detection/adapters/mmdet/config_utils.py]: Why don't we put this file into adapters/mmdet/utils folder? I think it's better to put all files in subfolders of mmdet. So make it remain no files in mmdet except for` __init__.py` like anomaly.

File: otx/algorithms/detection/adapters/mmdet/pipelines.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
+"""Collection Pipeline for detection task."""
```
Comment (supersoob) [inline otx/algorithms/detection/adapters/mmdet/pipelines.py]: Similar comments with above. I'd like to suggest put this in `data` folder together with dataset files.

File: otx/algorithms/detection/adapters/mmdet/dataset/mmdataset.py (hunk 10 above)
Comment (harimkang) [reply otx/algorithms/detection/adapters/mmdet/dataset/mmdataset.py]: I took a look at `anomalib `to see along with other comments below.
This has been modified to `mmdet/data/dataset.py`.

File: otx/algorithms/detection/adapters/mmdet/pipelines.py (hunk 12 above)
Comment (harimkang) [reply otx/algorithms/detection/adapters/mmdet/pipelines.py]: I moved it inside `mmdet/data`.

File: otx/algorithms/detection/adapters/mmdet/config_utils.py (hunk 11 above)
Comment (harimkang) [reply otx/algorithms/detection/adapters/mmdet/config_utils.py]: I moved it inside `mmdet/utils`. Thanks for the nice comment.

Comment (supersoob) [reply otx/algorithms/detection/adapters/mmdet/config_utils.py]: It looks great!

File: otx/algorithms/common/configs/base.py (hunk 5 above)
Comment (sungmanc) [reply otx/algorithms/common/configs/base.py]: I think it is not mandatory that all configurations defined under this file are consumed by all models even if it's named 'common'. If some configurations are frequently used for different tasks, then 'common' is the proper location.

So, how about using `common/configs/training_base.py` for these configurations?

**Comment by harimkang** [comment]
Hi code reviewers (with @ashwinvaidya17)
Currently, with the script that creates the `venv` in `otx/algorithms/detection/init_venv.sh`, it was confirmed that the otx cli & sample (with nncf, pot) run well locally, but with other installation methods (in tox), some version conflicts still occur.

so i would like to take a PR that adds a test code **separately**. (next PR)
in the next PR, I will add the unit-test and integration test and work on making it work in the otx environment.

Please review with reference to the above.

File: otx/algorithms/common/utils/data_utils.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions
+# and limitations under the License.
+
+from mpa.utils.logger import get_logger
+
+logger = get_logger()
```
Comment (eunwoosh) [inline otx/algorithms/common/utils/data_utils.py]: isn't it better to remove logger and "import get_logger", if not used?

File: otx/algorithms/common/utils/data_utils.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    for i, item in enumerate(dataset):
+        item_labels = item.annotation_scene.get_labels()
+        for i_l in item_labels:
+            if i_l in labels:
+                img_indices[i_l.name].append(i)
+
+    return img_indices
+
+
+def get_old_new_img_indices(labels, new_classes, dataset):
```
Comment (sungmanc) [inline otx/algorithms/common/utils/data_utils.py]: The reason why this method is located in `common/utils/data_utils.py` is our default learning method is `class_incremental`? I'm not sure that methods related to `class_incremental` are exposed on `common`.

Even if we selected `class_incremental` as a base, it is unfamiliar to developers who used a conventional framework.

File: otx/algorithms/detection/tasks/train.py [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+            "labels": labels,
+            "confidence_threshold": self.confidence_threshold,
+            "VERSION": 1,
+        }
+        if (
+            self._model_cfg
+            and hasattr(self._model_cfg.model, "bbox_head")
+            and hasattr(self._model_cfg.model.bbox_head, "anchor_generator")
+        ):
+            if getattr(
```
Comment (eunwoosh) [inline otx/algorithms/detection/tasks/train.py]: how about modifying getattr to hasattr and the line with the line right above it?

File: otx/algorithms/detection/tasks/train.py [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+        # update checkpoint to the newly trained model
+        self._model_ckpt = model_ckpt
+
+        # Update anchors
+        if (
+            self._model_cfg
+            and hasattr(self._model_cfg.model, "bbox_head")
+            and hasattr(self._model_cfg.model.bbox_head, "anchor_generator")
+        ):
+            if getattr(
```
Comment (eunwoosh) [inline otx/algorithms/detection/tasks/train.py]: how about modifying getattr to hasattr and the line with the line right above it?

File: otx/algorithms/detection/adapters/mmdet/config_utils.py (hunk 11 above)
Comment (harimkang) [reply otx/algorithms/detection/adapters/mmdet/config_utils.py]: Please resolve comments that have been resolved. :)

File: otx/algorithms/common/utils/data_utils.py (hunk 14 above)
Comment (harimkang) [reply otx/algorithms/common/utils/data_utils.py]: This is something I agree with to some extent. 
This needs to be cleaned up, but with the current structure, it is most efficient to proceed in the training task. 

I think this will also be included in the backlog of future work. Thanks for the good comments.

File: otx/algorithms/common/utils/data_utils.py (hunk 13 above)
Comment (harimkang) [reply otx/algorithms/common/utils/data_utils.py]: Thanks for the nice comment. Removed.

File: otx/algorithms/detection/tasks/train.py (hunk 15 above)
Comment (harimkang) [reply otx/algorithms/detection/tasks/train.py]: Edited.

File: otx/algorithms/detection/tasks/train.py (hunk 16 above)
Comment (harimkang) [reply otx/algorithms/detection/tasks/train.py]: Edited as above.

File: otx/algorithms/common/utils/data_utils.py (hunk 14 above)
Comment (sungmanc) [reply otx/algorithms/common/utils/data_utils.py]: I agree, let's remain some refactorings related to architecture.

File: otx/algorithms/detection/tasks/train.py (hunk 15 above)
Comment (eunwoosh) [reply otx/algorithms/detection/tasks/train.py]: I'm sorry I forgot the word combine. isn't it better to combine all hasattr lines together?

File: otx/algorithms/detection/tasks/train.py (hunk 16 above)
Comment (eunwoosh) [reply otx/algorithms/detection/tasks/train.py]: it's same as above

File: otx/algorithms/detection/utils/otx_utils.py [hunk 17]
```diff
@@ -0,0 +1,220 @@
+"""Utils for OTX Detection."""
```
Comment (ashwinvaidya17) [inline otx/algorithms/detection/utils/otx_utils.py]: Should we keep this just to `data.py` and the `otx_utils.py` to just `utils.py`? As they are in the utils folder the postfix can be dropped.

File: otx/algorithms/common/configs/base.py (hunk 5 above)
Comment (ashwinvaidya17) [reply otx/algorithms/common/configs/base.py]: That's true @sungmanc. Plus some anomaly models also require training parameters such as early stopping. I am fine with `common/configs/training_base.py`. Maybe others might also have some suggestions on the naming.

**Review by ashwinvaidya17 — APPROVED** [review]
Thanks I am fine with these changes. We can look into the tests in a different PR.

File: otx/algorithms/detection/tasks/train.py (hunk 15 above)
Comment (harimkang) [reply otx/algorithms/detection/tasks/train.py]: reflected.

File: otx/algorithms/detection/tasks/train.py (hunk 16 above)
Comment (harimkang) [reply otx/algorithms/detection/tasks/train.py]: reflected.

File: otx/algorithms/detection/utils/otx_utils.py (hunk 17 above)
Comment (harimkang) [reply otx/algorithms/detection/utils/otx_utils.py]: I agree with the comments and renamed this.

File: otx/algorithms/common/configs/base.py (hunk 5 above)
Comment (harimkang) [reply otx/algorithms/common/configs/base.py]: Renamed `base.py` in `common/configs` and `tasks` to `training_base.py`. Thanks for the nice comment.
